File: src/eavt_sql/query_codec.py
from __future__ import annotations

import struct
from datetime import datetime
from typing import Any

_TAG_TEXT = 1
_TAG_INT64 = 2
_TAG_FLOAT64 = 3
_TAG_BOOL = 4
_TAG_BYTES = 5
_TAG_TIMESTAMP = 6
_TAG_UNKNOWN = 99


def encode_values(values: list[Any]) -> bytes:
    buf = bytearray()
    buf += struct.pack(">I", len(values))
    for v in values:
        _encode_one(buf, v)
    return bytes(buf)


def decode_values(buf: bytes) -> list[Any]:
    if len(buf) < 4:
        raise ValueError("decode_values: buffer too short")
    n = struct.unpack_from(">I", buf, 0)[0]
    pos = 4
    out: list[Any] = []
    for _ in range(n):
        val, pos = _decode_one(buf, pos)
        out.append(val)
    return out


def _encode_one(buf: bytearray, v: Any) -> None:
    if isinstance(v, bool):
        buf.append(_TAG_BOOL)
        buf.append(1 if v else 0)
    elif isinstance(v, int):
        buf.append(_TAG_INT64)
        buf += struct.pack(">q", v)
    elif isinstance(v, float):
        buf.append(_TAG_FLOAT64)
        buf += struct.pack(">d", v)
    elif isinstance(v, str):
        _encode_bytes_payload(buf, _TAG_TEXT, v.encode("utf-8"))
    elif isinstance(v, (bytes, bytearray)):
        _encode_bytes_payload(buf, _TAG_BYTES, bytes(v))
    elif isinstance(v, datetime):
        buf.append(_TAG_TIMESTAMP)
        us = int(v.timestamp() * 1_000_000)
        buf += struct.pack(">q", us)
    elif v is None:
        buf.append(_TAG_INT64)
        buf += struct.pack(">q", 0)
    else:
        raise TypeError(f"unsupported value type: {type(v).__name__}")


def _encode_bytes_payload(buf: bytearray, tag: int, data: bytes) -> None:
    buf.append(tag)
    buf += struct.pack(">I", len(data))
    buf += data


def _decode_one(buf: bytes, pos: int) -> tuple[Any, int]:
    tag = buf[pos]
    pos += 1
    if tag == _TAG_TEXT:
        s, pos = _decode_bytes(buf, pos)
        return s.decode("utf-8"), pos
    elif tag == _TAG_INT64:
        n = struct.unpack_from(">q", buf, pos)[0]
        return n, pos + 8
    elif tag == _TAG_FLOAT64:
        f = struct.unpack_from(">d", buf, pos)[0]
        return f, pos + 8
    elif tag == _TAG_BOOL:
        return buf[pos] != 0, pos + 1
    elif tag == _TAG_BYTES:
        return _decode_bytes(buf, pos)
    elif tag == _TAG_TIMESTAMP:
        us = struct.unpack_from(">q", buf, pos)[0]
        return us, pos + 8
    elif tag == _TAG_UNKNOWN:
        t = buf[pos]
        payload = struct.unpack_from(">Q", buf, pos + 1)[0]
        return (t, payload), pos + 9
    else:
        raise ValueError(f"decode_one: unknown tag {tag}")


def _decode_bytes(buf: bytes, pos: int) -> tuple[bytes, int]:
    length = struct.unpack_from(">I", buf, pos)[0]
    start = pos + 4
    return buf[start : start + length], start + length


def decode_rows(buf: bytes) -> list[tuple]:
    """Decode cursor batch format: [u32 num_cols][encoded values]... per row.

    Rows with num_cols == 0 are skipped (internal VM bookkeeping).
    """
    rows: list[tuple] = []
    pos = 0
    while pos < len(buf):
        num_cols = struct.unpack_from(">I", buf, pos)[0]
        pos += 4
        row: list[Any] = []
        for _ in range(num_cols):
            val, pos = _decode_one(buf, pos)
            row.append(val)
        if num_cols > 0:
            rows.append(tuple(row))
    return rows

File: dynspire-commons/src/transactor/query_codec.rs
use super::Value;

/// Serialize a flat `Vec<Value>` to bytes.
///
/// Format:
/// ```text
/// [u32 BE num_values]
/// for each value:
///   [u8 tag]
///   [tag-specific data]
/// ```
///
/// Tags: 0=Null, 1=Text, 2=Int64, 3=Float64, 4=Bool, 5=Bytes, 6=Timestamp, 7=EntityId
pub fn encode_values(values: &[Value]) -> Vec<u8> {
    let mut buf = Vec::with_capacity(values.len() * 12 + 4);
    buf.extend_from_slice(&(values.len() as u32).to_be_bytes());
    for v in values {
        encode_one(&mut buf, v);
    }
    buf
}

/// Deserialize a flat `Vec<Value>` from bytes produced by [`encode_values`].
pub fn decode_values(buf: &[u8]) -> Result<Vec<Value>, String> {
    if buf.len() < 4 {
        return Err("decode_values: buffer too short".into());
    }
    let n = u32::from_be_bytes([buf[0], buf[1], buf[2], buf[3]]) as usize;
    let mut pos = 4;
    let mut out = Vec::with_capacity(n);
    for _ in 0..n {
        let (v, np) = decode_one(buf, pos)?;
        out.push(v);
        pos = np;
    }
    Ok(out)
}

/// Decode cursor batch format: `[u32 ncols][values]...` per row.
/// Returns rows as `Vec<Vec<Value>>`. Rows with ncols==0 are skipped.
pub fn decode_rows(buf: &[u8]) -> Result<Vec<Vec<Value>>, String> {
    let mut rows = Vec::new();
    let mut pos = 0;
    while pos < buf.len() {
        if pos + 4 > buf.len() {
            return Err("decode_rows: truncated ncols".into());
        }
        let ncols = u32::from_be_bytes([buf[pos], buf[pos + 1], buf[pos + 2], buf[pos + 3]]) as usize;
        pos += 4;
        let mut row = Vec::with_capacity(ncols);
        for _ in 0..ncols {
            let (v, np) = decode_one(buf, pos)?;
            row.push(v);
            pos = np;
        }
        if ncols > 0 {
            rows.push(row);
        }
    }
    Ok(rows)
}

pub fn encode_one(buf: &mut Vec<u8>, v: &Value) {
    match v {
        Value::Text(s) => {
            buf.push(1);
            let bytes = s.as_bytes();
            buf.extend_from_slice(&(bytes.len() as u32).to_be_bytes());
            buf.extend_from_slice(bytes);
        }
        Value::Int64(n) => {
            buf.push(2);
            buf.extend_from_slice(&n.to_be_bytes());
        }
        Value::Float64(f) => {
            buf.push(3);
            buf.extend_from_slice(&f.to_be_bytes());
        }
        Value::Bool(b) => {
            buf.push(4);
            buf.push(*b);
        }
        Value::Bytes(b) => {
            buf.push(5);
            buf.extend_from_slice(&(b.len() as u32).to_be_bytes());
            buf.extend_from_slice(b);
        }
        Value::Timestamp(us) => {
            buf.push(6);
            buf.extend_from_slice(&us.to_be_bytes());
        }
        Value::Unknown(tag, payload) => {
            buf.push(99);
            buf.push(*tag as u8);
            buf.extend_from_slice(&payload.to_be_bytes());
        }
    }
}

fn decode_one(buf: &[u8], pos: usize) -> Result<(Value, usize), String> {
    if pos >= buf.len() {
        return Err("decode_one: unexpected end".into());
    }
    let tag = buf[pos];
    let pos = pos + 1;
    match tag {
        1 => {
            let (s, np) = read_str(buf, pos)?;
            Ok((Value::Text(s), np))
        }
        2 => {
            if pos + 8 > buf.len() {
                return Err("decode_one: truncated Int64".into());
            }
            let n = i64::from_be_bytes(buf[pos..pos + 8].try_into().unwrap());
            Ok((Value::Int64(n), pos + 8))
        }
        3 => {
            if pos + 8 > buf.len() {
                return Err("decode_one: truncated Float64".into());
            }
            let f = f64::from_be_bytes(buf[pos..pos + 8].try_into().unwrap());
            Ok((Value::Float64(f), pos + 8))
        }
        4 => {
            if pos >= buf.len() {
                return Err("decode_one: truncated Bool".into());
            }
            Ok((Value::Bool(buf[pos]), pos + 1))
        }
        5 => {
            let (b, np) = read_bytes(buf, pos)?;
            Ok((Value::Bytes(b), np))
        }
        6 => {
            if pos + 8 > buf.len() {
                return Err("decode_one: truncated Timestamp".into());
            }
            let us = i64::from_be_bytes(buf[pos..pos + 8].try_into().unwrap());
            Ok((Value::Timestamp(us), pos + 8))
        }
        99 => {
            if pos + 1 + 8 > buf.len() {
                return Err("decode_one: truncated Unknown".into());
            }
            let tag = buf[pos] as i8;
            let payload = u64::from_be_bytes(buf[pos + 1..pos + 9].try_into().unwrap());
            Ok((Value::Unknown(tag, payload), pos + 9))
        }
        _ => Err(format!("decode_one: unknown tag {tag}")),
    }
}

fn read_str(buf: &[u8], pos: usize) -> Result<(String, usize), String> {
    let (bytes, np) = read_bytes(buf, pos)?;
    let s = String::from_utf8(bytes).map_err(|e| format!("invalid utf8: {e}"))?;
    Ok((s, np))
}

fn read_bytes(buf: &[u8], pos: usize) -> Result<(Vec<u8>, usize), String> {
    if pos + 4 > buf.len() {
        return Err("read_bytes: truncated length".into());
    }
    let len = u32::from_be_bytes([buf[pos], buf[pos + 1], buf[pos + 2], buf[pos + 3]]) as usize;
    let start = pos + 4;
    if start + len > buf.len() {
        return Err("read_bytes: truncated data".into());
    }
    Ok((buf[start..start + len].to_vec(), start + len))
}
